_source/_posts/2021-12-03-angular-dynamic-components.md

```diff
@@ -0,0 +1,773 @@
+      map((s: AuthState) => +s.idToken?.claims['department'] ?? 0),  
+      take(1)  
+    ).subscribe(  
+      (d: number) => this.messages = this._messageService.getMessages(d)  
+    );  
+  }
+}
+```
+
+We now have a list of messages to display to the user. We'll pass this to the dynamic component's container component to action. In the inline template for `ProtectedComponent`, update the `<app-department>` element to pass in `messages` as an input property. You'll see an error in the IDE since we haven't created the input property in the Department component yet. The element in the inline template will look like the code below.
+
+{% raw %}
+```html
+<app-department [messages]="messages"></app-department>
+```
+{% endraw %}
+
+You can probably see where this is going. Now we'll update the Department component to support the input property. Open `src/app/protected/department/department.component.ts`. Replace the hardcoded private `messages` property into a public input property like the code snippet below.
+
+```ts
+@Input() public messages: MessageItem[] = [];
+```
+
+You'll have a new error in the app because our code in the `loadComponent` method expects a component type. We'll add a factory method to return the component type to create by matching the `MessageType` to the component type like the following example.
+
+{% raw %}
+```ts
+private componentTypeFactory(type: MessageType): Type<DynamicComponent> {  
+  let comp: Type<DynamicComponent>;  
+  if (type === 'Pawesome') {  
+    comp = PawesomeComponent;  
+  } else if (type === 'Clawesome') {  
+    comp = ClawesomeComponent;  
+  } else {  
+    comp = SmileyComponent;  
+  }  
+ return comp;  
+}
+```
+{% endraw %}
+
+Then we can update the `loadComponent` method to use the factory method. We also have message data to pass into the components, although the dynamic components can't support the input property. Update the code and add the new line of code to pass data to the components like the code block below.
+
+{% raw %}
+```ts
+const componentRef = viewContainerRef.createComponent<DynamicComponent>(this.componentTypeFactory(message.type));  
+componentRef.instance.data = message.data;
+```
+{% endraw %}
+
+Now everything should compile, and you have a working app that displays a rotating set of cute animals images to help you power through working on that task list. Try logging in as a user with a different department (or change the department value for yourself in the Okta dashboard) to see the various dynamic components at work.
+
+You can find the code for [this project on GitHub](https://github.com/oktadev/okta-angular-dynamic-components-example).
+
+We did a lot in this tutorial and had a lot of dependencies on Material and Okta libraries. Please check out [this commit](https://github.com/oktadev/okta-angular-dynamic-components-example/commit/3f1e18092e30ee17c8722b73899a52adb825be50) to see the changes required to run tests and mock the Okta services.
```

Comment:
```suggestion
We did a lot in this tutorial and had a lot of dependencies on Material and Okta libraries. Please check out [this commit](https://github.com/oktadev/okta-angular-dynamic-components-example/commit/3f1e18092e30ee17c8722b73899a52adb825be50) to see the changes required to run tests and mock up the Okta services.
```
removes ambiguity

Reply:
@havi I'd like to change this to one of the following instead since "mock up" doesn't feel right for a testing mock. Thoughts on which one?

> ... to see the changes required to run tests and mock out the Okta services.
or
> ... to see the changes required to run tests and create mocks of the Okta services.